### PR TITLE
feat: return generated keys

### PR DESCRIPTION
**Pull request by olavloite**
Adds support for returning generated keys.

### CODE REVIEW

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcConnection.java [hunk 1]
```diff
@@ -299,22 +309,26 @@ public PreparedStatement prepareStatement(
...
+            : NO_GENERATED_KEY_COLUMNS);
   }
 
   @Override
   public PreparedStatement prepareStatement(String sql, int[] columnIndexes) throws SQLException {
-    checkClosed();
-    return prepareStatement(sql);
+    // This should preferably have returned an error, but the initial version of the driver just
+    // accepted and ignored this. Starting to throw an error now would be a breaking change.
+    // TODO: Consider throwing an Unsupported error for the next major version bump.
```
Comment (rajatbhatta) [inline src/main/java/com/google/cloud/spanner/jdbc/JdbcConnection.java:323]: ACK.

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java [hunk 2]
```diff
@@ -441,77 +499,94 @@ void convertUpdateCountsToSuccessNoInfo(long[] updateCounts, long[] res) {
...
+            : NO_GENERATED_KEY_COLUMNS);
   }
 
   @Override
   public long executeLargeUpdate(String sql, int[] columnIndexes) throws SQLException {
-    checkClosed();
-    return executeLargeUpdate(sql);
+    // This should preferably have returned an error, but the initial version of the driver just
+    // accepted and ignored this. Starting to throw an error now would be a breaking change.
+    // TODO: Consider throwing an Unsupported error for the next major version bump.
```
Comment (rajatbhatta) [inline src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java:554]: ACK.

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java [hunk 3]
```diff
@@ -441,77 +499,94 @@ void convertUpdateCountsToSuccessNoInfo(long[] updateCounts, long[] res) {
...
+            : NO_GENERATED_KEY_COLUMNS);
   }
 
   @Override
   public int executeUpdate(String sql, int[] columnIndexes) throws SQLException {
-    checkClosed();
-    return executeUpdate(sql);
+    // This should preferably have returned an error, but the initial version of the driver just
+    // accepted and ignored this. Starting to throw an error now would be a breaking change.
+    // TODO: Consider throwing an Unsupported error for the next major version bump.
```
Comment (rajatbhatta) [inline src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java:530]: ACK.

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java [hunk 4]
```diff
@@ -105,6 +128,29 @@ public long executeLargeUpdate(String sql) throws SQLException {
     }
   }
 
+  /** Extracts the update count from the given result set and then closes the result set. */
+  private long extractUpdateCountAndClose(com.google.cloud.spanner.ResultSet resultSet)
```
Comment (rajatbhatta) [inline src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java:132]: Will we be able to extract the stats without consuming the rows? IIRC, `resultSet.getStats()` can be accessed only when `resultSet.next()` returns `false` (no more rows to consume).

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java [hunk 5]
```diff
@@ -105,6 +128,29 @@ public long executeLargeUpdate(String sql) throws SQLException {
...
+      throws SQLException {
+    try {
+      if (resultSet.getStats() == null) {
+        throw JdbcSqlExceptionFactory.of(
+            "Result does not contain any stats", Code.FAILED_PRECONDITION);
+      }
+      long updateCount;
+      if (resultSet.getStats().hasRowCountExact()) {
+        updateCount = resultSet.getStats().getRowCountExact();
+      } else if (resultSet.getStats().hasRowCountLowerBound()) {
```
Comment (rajatbhatta) [inline src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java:142]: Will this lower bound be the correct `updateCount` value?

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java (hunk 4 above)
Comment (olavloite) [reply src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java:139]: Yes, I've added some additional comments to clarify. We only call this method with a `ResultSet` that has already been fully consumed. That happens in the `copyOf` method.

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java [hunk 6]
```diff
@@ -87,14 +94,33 @@ public int executeUpdate(String sql) throws SQLException {
...
       case RESULT_SET:
-        // Close the result set as we are not going to return it to the user. This prevents the
-        // underlying session from potentially being leaked.
-        throw closeResultSetAndCreateInvalidQueryException(result);
+        if (generatedKeysColumns.isEmpty()) {
+          // Close the result set as we are not going to return it to the user. This prevents the
+          // underlying session from potentially being leaked.
+          throw closeResultSetAndCreateInvalidQueryException(result);
+        }
+        // Make a copy of the result set os it does not matter if the user does not close the result
```
Comment (rajatbhatta) [inline src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java:119]:
```suggestion
        // Make a copy of the result set as it does not matter if the user does not close the result
```

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java (hunk 4 above)
Comment (rajatbhatta) [reply src/main/java/com/google/cloud/spanner/jdbc/JdbcStatement.java:139]: Makes sense.

**Review by rajatbhatta — APPROVED** [review]

**Pull request merged**